File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder sb = new StringBuilder();

        int n = Integer.parseInt(br.readLine());
        for (int i = 0; i < n; i++) {
            Stack<Character> stack = new Stack<>();
            String input = br.readLine();

            for (int j = 0; j < input.length(); j++) {
                if (input.charAt(j) == '(') {
                    stack.push(input.charAt(j));
                } else if (input.charAt(j) == ')') {
                    if (!stack.isEmpty() && stack.peek() == '(') {
                        stack.pop();
                    } else {
                        stack.push(input.charAt(j));
                    }
                }
            }

            if (stack.isEmpty()) {
                sb.append("YES").append('\n');
            } else {
                sb.append("NO").append('\n');
            }
        }

        bw.write(sb.toString());
        bw.flush();
        bw.close();
    }
}

File: 백준/Silver/9012. 괄호/괄호.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val sb = StringBuilder()

    val n = br.readLine().toInt()
    repeat(n) {
        val stack = Stack<Char>()
        val input = br.readLine()
        for(c in input.toCharArray()) {
            if (c == '(') {
                stack.push(c)
            }else if (c == ')') {
                if (stack.isEmpty()) {
                    stack.push(c)
                } else if (stack.peek() == '(') {
                    stack.pop()
                }
            }
        }

        if (stack.isEmpty()) {
            sb.append("YES").append("\n")
        } else {
            sb.append("NO").append("\n")
        }
    }

    bw.write(sb.toString())
    bw.flush()
    bw.close()
}